### ISSUE TITLE
Support communication-efficient optimizers

### ISSUE THREAD

**Issue by ploshkin** [learning-at-home/hivemind#184]
`<edit>`
The goal of this project is to implement pre-configured decentralized optimizers such as [DADAM](https://arxiv.org/abs/1901.09109) and decentralized [Local AMSGrad](https://dl.acm.org/doi/10.1145/3412815.3416891). We intend to create a general interface that runs adaptive optimizers with periodic averaging of gradients, parameters and optimizer statistics on different time-frames]

The final version should support:
* [ ] wrapping any adam-like optimizers such as Adam, AMSGrad, LAMB(FusedLAMB), Radam, etc
* [ ] full torch.optim.Optimizer interface including state_dict/load_state_dict
   * [ ] for parameter groups, either support it or explicitly throw an error
* [ ] averaging optimizer statistics in background while the model is training, incorporating local updates after the averaging is done
* [ ] have a default confguration that converges :)
  * [ ] test convergence locally with a toy task
  * [ ] test in standard ALBERT-base setup for wikitext103
  * [ ] test with ALBERT-large on OpenBookCorpus

`</edit>`

Code draft:

```
# usage:
class LAMBWithAveraging(FusedLAMB, AdaptiveOptimizer):
    def get_averaged_statistics() -> Sequence[Tensor]
       return TODO()

opt = DecentralizedOptimizer(model, base_optimizer=LAMBWithAveraging(model.parameters()),
                             prefix='MY_UNIQUE_STRING')


### code

class AdaptiveOptimizer(abc.Mixin):
   def get_averaged_statistics(self):
     raise NotImplementedError()

class DecentralizedOptimizer(abc.meta):
    def __init__(self, parameters, base_optimizer, dht, prefix: str, ..., ):
       assert isinstance(base_optimizer, AdaptiveOptimizer), "TODO see example here"

       self.parameters = list(parameters)  # on GPU
       self.base_optimizer = base_optimizer
       self.global_step = ...
       
       zeros_like_grads = [torch.zeros_like(x.detach().cpu()) for x in self.parameters]
       self.grad_averager = DecentralizedAverager(
           averaged_tensors= zeros_like_grads  # average gradients only
            prefix=prefix+'_grad_averager', **some_kwargs
           )
       
       parameters = [x.cpu().detach().float() self.parameters()]
       averaged_statistics = [x.cpu().detach().float() for x in base_optimizer.get_averaged_statistics()]
       
       self.almost_v_averager = DecentralizedAverager(
           averaged_tensors= parameters + averaged_statistics  # average params and second momenta
            prefix=prefix+'_v_averager', **some_other_kwargs
           )
       self.v_averaging_job = None
       
    def get_model_params_and_v(self) -> Tuple[torch.Tensor, ...]:
        TODO()
              
    def step(self):
       accumulate_grads()
       report_progress_to_dht()
       
       if self.v_averaging_job.done():
          local_params_and_v = self.paramters + self.base_optimzier.get_averaged_stats()
          with self.almost_v_averager.get_tensors() as averaged_params_and_v:
             copy_from(averaged_params_and_v).to(local_params_and_v)

       
       if accumulated_enough():
       
           local_grads = [x.grad for x in model.parameters()]
           with self.grad_averager.get_tensors() as averaged_grads:
               copy_from(local_grads).to(averaged_grads)
           self.grad_averager.step()
           with self.grad_averager.get_tensors() as averaged_grads:
               copy_from(averaged_grads).to(local_grads)
           # TODO maybe average gradients continuously in background
             
           if not self.use_local_statistics:
               TODO(revert base_optimizer.get_averaged_statistics() to prev averaged state)
           self.base_optimizer.step()
           self.global_step += 1  # MAKE SURE IT IS in sync with collaboration
           
           if not self.use_local_statistics:
               TODO(revert optimizer statistics back to local ones AFTER local update)
           
           
           ### AVERAGE PARAMETERS AND STATISTICS! ###
           if self.global_step % self.K == 0:
               if not v_averaging_job.done():
                   if self.await_v_averager:
                      v_averaging_job.wait()
                      local_params_and_v = self.paramters + self.base_optimzier.get_averaged_stats()
                      with self.almost_v_averager.get_tensors() as averaged_params_and_v:
                         copy_from(averaged_params_and_v).to(local_params_and_v)
            
                   else:
                       warning("Param averager is taking too long, TODO explain why we skip")
                       return
           
               local_params_and_v = self.paramters + self.base_optimzier.get_averaged_stats()
               with self.almost_v_averager.get_tensors() as averaged_params_and_v:
                   copy_from(local_params_and_v).to(averaged_params_and_v)
               
               self.v_averaging_job = run_in_background(almost_v_averager.step())
           
       
    def zero_grad(self):
       self.opt.zero_grad()
       zero_grads_in_self_grad_averager()
       self.report_to_dht()
```

**Comment by justheuristic** [comment]
/* echoing what we agreed upon 1h ago in telegram */
* the first version can be tested locally with torch.optim.Adam
* better *not* to have AdaptiveOptimizer, but instead pass some kind of `averaged_suffixes` that selects all statistics from opt.state_dict
* would be great to mimic the interface and arg names from [fairscale OSS](https://github.com/facebookresearch/fairscale)
* would be great to inherit DecentralizedOptimizer from Optimizer and create dummy state_dict/load_state_dict/add_param_grou/param_groups
*  for now they can simply run NotImplemented error

__Note:__ to fully initialize optimizer statistics, run:
```
for param in parameters:
    (param.sum() * 0) .backward()
opt.step()
```
Also, please check that the number of averaged statistics does NOT change between optimizer steps

**Comment by justheuristic** [comment]
Update after #215 -
- please implement the statistic averaging logic in [TrainingAverager](https://github.com/learning-at-home/hivemind/blob/master/hivemind/client/averaging/training.py)?
- we can create a separate AdaptiveOptimizer (or any name you prefer) that is similar to CollaborativeOptimizer, but also accounts for adaptive learning rates (and NOT gradients :).

If we find that it is strictly better than current CollaborativeOptimizer, we can switch entirely to the new version and deprecate the original.
